### PR TITLE
KAFKA-10586, KAFKA-9981: Intra-cluster communication for Mirror Maker 2

### PR DESCRIPTION
**Pull request by C0urante**
Relevant Jira tickets:
- [KAFKA-10586](https://issues.apache.org/jira/browse/KAFKA-10586)
- [KAFKA-9981](https://issues.apache.org/jira/browse/KAFKA-9981)

Implements the internal REST API changes described in [KIP-710](https://cwiki.apache.org/confluence/display/KAFKA/KIP-710%3A+Full+support+for+distributed+mode+in+dedicated+MirrorMaker+2.0+clusters).

A lot of the diff here comes from shuffling existing parts of the code base into new parts, without significantly altering them. This is necessary in order to make the current REST API logic for Kafka Connect more extensible and reusable, which in turn allows us to add logic for the dedicated Mirror Maker 2 REST API with less effort.

Shuffled but not extensively rewritten portions include:
- `WorkerConfig` properties being extracted into the new `RestServerConfig` class
- Pulling out some REST forwarding logic previously internal to the `ConnectorsResource` class into the new `HerderRequestHandler` class
- Pulling out all REST logic for internal endpoints from the `ConnectorsResource` class into the new `InternalClusterResource` class

These changes should not affect the docs generated by, e.g., the `DistributedConfig` class, but should allow other `AbstractConfig` classes to easily add properties related to the (public- or internal-facing) Connect REST API.

An integration test is also added that runs a multi-node, dedicated Mirror Maker 2 cluster with exactly-once support enabled in order to test out both code paths related to intra-cluster communication.

### CODE REVIEW

**Comment by C0urante** [comment]
@gharris1727 @viktorsomogyi @mimaison would you mind taking a look at this when you have a moment?

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java
Comment (gharris1727) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java]: interesting, the for-testing constructor appears unused, and has been for as long as it's existed (ever since #321).
We can push this refactor out to a different PR.

**Review by gharris1727 — APPROVED** [review]
LGTM!

Thanks for refactoring the resource hierarchy and making the RestServer abstract.
I think that makes a lot of sense given the different contexts for the RestServer in MM2 vs the other modes.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java
Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java]: Blegh, this comment was left in from an earlier draft where there was an additional constructor. I refactored it out before opening the PR, but apparently forgot to remove the comment. Took it out now, thanks for catching!

File: connect/mirror/src/test/java/org/apache/kafka/connect/mirror/integration/DedicatedMirrorIntegrationTest.java
Comment (urbandan) [inline connect/mirror/src/test/java/org/apache/kafka/connect/mirror/integration/DedicatedMirrorIntegrationTest.java:163]: Can you elaborate on how this will ensure that the task config update works with the inter-node comm?
Is there a way to ensure that the MirrorSourceConnector will be assigned to a follower?
Just wondering if this test can pass most of the time, and will flakily fail 1 out of X if there is an issue with the task config update.

File: connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorMaker.java
Comment (viktorsomogyi) [inline connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorMaker.java:135]: What prevents this from being null or Optional<String>?

File: connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorMaker.java
Comment (viktorsomogyi) [inline connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorMaker.java:306]: Is this a Jetty bug or just some loophole in URL encoding in general?

**Review by viktorsomogyi — COMMENTED** [review]
Left a few small comments but will continue the review tomorrow.

File: connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorMaker.java
Comment (C0urante) [reply connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorMaker.java:306]: It's the result of fitting a square peg (Java's `URLEncoder` class, which, despite the name, is designed for HTML form encoding instead of URL path encoding) into a round hole (URL path encoding). I couldn't find a better alternative than this, and considering the fairly low risk (this is all intended to cover a pretty niche edge case) and integration test coverage, figured it'd be good enough for now.

File: connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorMaker.java
Comment (C0urante) [reply connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorMaker.java:135]: It ends up being used in the metadata workers send to the group coordinator during rebalances, and the schema we use to (de)serialize that data requires a non-null string for the worker URL. I stuck with the `"NOTUSED"` placeholder instead of an empty string since that's what we're using currently, and it doesn't seem worth the risk to try to change it now.

File: connect/mirror/src/test/java/org/apache/kafka/connect/mirror/integration/DedicatedMirrorIntegrationTest.java
Comment (C0urante) [reply connect/mirror/src/test/java/org/apache/kafka/connect/mirror/integration/DedicatedMirrorIntegrationTest.java:163]: We don't have a guarantee that task configs are relayed from one worker to another, that's true. However, because we enable exactly-once support, we do get a guarantee that intra-cluster communication takes place, since no source tasks can start on a follower node without first issuing a REST request to the cluster's leader.

Comment (urbandan) [reply connect/mirror/src/test/java/org/apache/kafka/connect/mirror/integration/DedicatedMirrorIntegrationTest.java:163]: Thanks for the clarification. I think using the EOS endpoints should be enough, as I don't really have any ideas on how to make sure that the task config update is triggered.

File: connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorMaker.java
Comment (viktorsomogyi) [reply connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorMaker.java:306]: Ok, I'm fine with it.

**Review by viktorsomogyi — COMMENTED** [review]
@C0urante I've finished the review and didn't find anything else to nitpick, but since I'm not that familiar with the details of Connect as you are, I would wait for a review from @mimaison. Or if you feel confident enough, you can submit it after resolving the conflicts :).

**Review by viktorsomogyi — APPROVED** [review]

File: connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorMakerConfig.java
Comment (mimaison) [inline connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorMakerConfig.java]: Nit: We usually have a period at the end of descriptions.

File: connect/mirror/src/main/java/org/apache/kafka/connect/mirror/rest/resources/InternalMirrorResource.java
Comment (mimaison) [inline connect/mirror/src/main/java/org/apache/kafka/connect/mirror/rest/resources/InternalMirrorResource.java]: Should we add a message here?

File: connect/mirror/src/main/java/org/apache/kafka/connect/mirror/rest/resources/InternalMirrorResource.java
Comment (mimaison) [inline connect/mirror/src/main/java/org/apache/kafka/connect/mirror/rest/resources/InternalMirrorResource.java]: Since we throw below, do we need this debug line?

File: connect/mirror/src/main/java/org/apache/kafka/connect/mirror/rest/resources/InternalMirrorResource.java
Comment (mimaison) [inline connect/mirror/src/main/java/org/apache/kafka/connect/mirror/rest/resources/InternalMirrorResource.java]: Could we use `getFirst()` and not have to check whether result is empty or do `get(0)`.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/RestServer.java
Comment (mimaison) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/RestServer.java]: nit: slightly more readable without `this.` here

**Review by mimaison — COMMENTED** [review]
Thanks for the PR, it looks great! I left a few very minor suggestions and it also needs another rebase.

File: connect/mirror/src/main/java/org/apache/kafka/connect/mirror/rest/resources/InternalMirrorResource.java
Comment (C0urante) [reply connect/mirror/src/main/java/org/apache/kafka/connect/mirror/rest/resources/InternalMirrorResource.java]: Yeah, couldn't hurt. This is almost guaranteed to end up in a log file on another worker so in the unlikely event that this code path is somehow reached, more detail would be preferable to less.

File: connect/mirror/src/main/java/org/apache/kafka/connect/mirror/rest/resources/InternalMirrorResource.java
Comment (C0urante) [reply connect/mirror/src/main/java/org/apache/kafka/connect/mirror/rest/resources/InternalMirrorResource.java]: I was under the impression that throwing this would only cause the exception message to be present in the REST response (and presumably the logs for the worker that issued the request), but after revisiting the `ConnectExceptionMapper` class, it appears that we'll also log the exception at debug level on this worker. So yep, we can remove this 👍

**Comment by C0urante** [comment]
Thanks Mickael, I've addressed your suggestions and rebased.

**Review by mimaison — APPROVED** [review]
LGTM, thanks @C0urante

**Comment by C0urante** [comment]
Thanks for the reviews, everyone!

**Comment by C0urante** [comment]
Whoops--there was a typo in the title at the time of merge. It should refer to [KAFKA-10586](https://issues.apache.org/jira/browse/KAFKA-10586), not KAFKA-15086.

**Pull request merged**
